crates/api/src/id.rs

```diff
@@ -0,0 +1,288 @@
+    /// Returns false if the default was unable to be set.
+    pub fn set_global_display_callback(cb: DisplayCb) -> bool {
+        AGENT_DISP.set(cb).is_ok()
+    }
+}
+
+static SPACE_DISP: std::sync::OnceLock<DisplayCb> = std::sync::OnceLock::new();
+
+/// Identifies a space to be tracked by Kitsune.
+#[derive(
+    Clone,
+    PartialEq,
+    Eq,
+    PartialOrd,
+    Ord,
+    Hash,
+    serde::Serialize,
+    serde::Deserialize,
+)]
+pub struct SpaceId(pub Id);
+
+imp_deref!(SpaceId, Id);
+imp_from!(SpaceId, bytes::Bytes, b => SpaceId(Id(b)));
+imp_from!(SpaceId, Id, b => SpaceId(b));
+
+impl std::fmt::Display for SpaceId {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        display(&self.0 .0, f, &SPACE_DISP)
+    }
+}
+
+impl std::fmt::Debug for SpaceId {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        display(&self.0 .0, f, &SPACE_DISP)
+    }
+}
+
+impl SpaceId {
+    /// Set the display/debug implementation for SpaceId for the duration
+    /// of this process. Note, if anything was printed earlier, the
+    /// default impl will have been set and cannot be changed.
+    /// Returns false if the default was unable to be set.
+    pub fn set_global_display_callback(cb: DisplayCb) -> bool {
+        SPACE_DISP.set(cb).is_ok()
+    }
+}
+
+static OP_DISP: std::sync::OnceLock<DisplayCb> = std::sync::OnceLock::new();
+
+/// Identifies an op to be tracked by Kitsune.
+#[derive(
+    Clone,
+    PartialEq,
+    Eq,
+    PartialOrd,
+    Ord,
+    Hash,
+    serde::Serialize,
+    serde::Deserialize,
+)]
+pub struct OpId(pub Id);
+
+imp_deref!(OpId, Id);
+imp_from!(OpId, bytes::Bytes, b => OpId(Id(b)));
+imp_from!(OpId, Id, b => OpId(b));
+
+impl std::fmt::Display for OpId {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        display(&self.0 .0, f, &OP_DISP)
+    }
+}
+
+impl std::fmt::Debug for OpId {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        display(&self.0 .0, f, &OP_DISP)
+    }
+}
+
+impl OpId {
+    /// Set the display/debug implementation for OpId for the duration
+    /// of this process. Note, if anything was printed earlier, the
+    /// default impl will have been set and cannot be changed.
+    /// Returns false if the default was unable to be set.
+    pub fn set_global_display_callback(cb: DisplayCb) -> bool {
+        OP_DISP.set(cb).is_ok()
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    #[test]
+    fn loc_zero_is_zero() {
+        assert_eq!(0, Id(bytes::Bytes::from_static(b"")).loc());
+    }
+
+    #[test]
+    fn loc_u32_equality() {
+        for u in [0, 42, 999, u32::MAX / 13, u32::MAX / 4, u32::MAX] {
+            assert_eq!(
+                u,
+                Id(bytes::Bytes::copy_from_slice(&u.to_le_bytes())).loc()
+            );
+        }
+    }
+
+    #[test]
+    fn loc_fixtures() {
+        const F: &[(&[u8], u32)] = &[
```

Comment:
Looks like a [test-case](https://crates.io/crates/test-case)

Reply:
We agreed to go without this dep for now, but we can always choose differently in the future.